### PR TITLE
tests: use "cat" also on Windows

### LINKED ISSUE #10574: tests: use shell-test / other helper instead of cat
There are a lot of tests being skipped still on Windows due to lack of `cat -`, or handled with something different (`find`, via https://github.com/neovim/neovim/commit/03e69a5d9c91b8fd709b8992c862b51520f60c98).

This should use a consistent helper/method, e.g. a new mode with shell-test, which would behave like `cat` then.

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/neovim/neovim/issues/10574.

Left out https://github.com/neovim/neovim/blob/1e2af5e39da9a5d0cc0e6940de343ef3f70cd8d0/test/functional/eval/system_spec.lua#L308.

### CODE REVIEW

File: test/functional/core/job_spec.lua [hunk 1]
```diff
@@ -759,7 +753,7 @@ describe('jobs', function()
   it('cannot have both rpc and pty options', function()
     command("let g:job_opts.pty = v:true")
     command("let g:job_opts.rpc = v:true")
-    local _, err = pcall(command, "let j = jobstart(has('win32') ? ['find', '/v', ''] : ['cat', '-'], g:job_opts)")
+    local _, err = pcall(command, "let j = jobstart(['cat', '-'], g:job_opts)")
```
Comment (blueyed) [inline test/functional/core/job_spec.lua]: Here any command should work probably anyway.

File: test/functional/core/job_spec.lua [hunk 2]
```diff
@@ -273,7 +268,7 @@ describe('jobs', function()
   end)
 
   it('disallows jobstop twice on the same job', function()
-    nvim('command', "let j = jobstart(has('win32') ? ['find', '/v', ''] : ['cat', '-'], g:job_opts)")
+    nvim('command', "let j = jobstart(['cat', '-'], g:job_opts)")
```
Comment (justinmk) [inline test/functional/core/job_spec.lua]: Removing these "find" workarounds is likely to fail. We already tried cat.exe for these cases and there were inconsistencies on windows/appveyor.

File: test/functional/core/job_spec.lua (hunk 1 above)
Comment (justinmk) [reply test/functional/core/job_spec.lua]: At long as the command doesn't exit quickly. That's why `cat -` is used: to make the process long-lived.

Comment (blueyed) [reply test/functional/core/job_spec.lua]: It is checked before the job is started (IIRC), or should be at least.

File: test/functional/core/job_spec.lua (hunk 2 above)
Comment (blueyed) [reply test/functional/core/job_spec.lua]: There is only one real failure/difference it seems (see fb43902).

**Comment by blueyed** [comment]
Via https://github.com/neovim/neovim/commit/03e69a5d9c91b8fd709b8992c862b51520f60c98#r34403873:

> It looks promising, except for one test, which might actually show an indifference (10 msgs on Linux, 5 on Windows): "jobs jobwait can be called recursively".
Gave it a last try on AppVeyor - but it would certainly be easier for you (or @erw7, or somebody having Windows locally) to investigate further.
The test was previously disabled, not using `find`.

**Comment by erw7** [comment]
The reason why there is missing "closed" msg is that it used the powershell alias cat. It ended with an error message like this:

```
cat : Cannot find path 'C:\Users\erw7\src\neovim\build\-' because it does not exist.
```

Therefore, it will work with the following modifications. 

```diff
--- a/test/functional/helpers.lua
+++ b/test/functional/helpers.lua
@@ -496,7 +496,7 @@ end
 local function set_shell_powershell()
   source([[
     set shell=powershell shellquote=( shellpipe=\| shellredir=> shellxquote=
-    let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command Remove-Item -Force alias:sleep;'
+    let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command Remove-Item -Force alias:sleep; Remove-Item -Force alias:cat;'
   ]])
 end
```

https://github.com/neovim/neovim/pull/10580/commits/fb43902713b892d9339880a3241d131b5b46e56e, https://github.com/neovim/neovim/pull/10580/commits/d6db1fa74b5af8c4d0819beb357030bac6d72e83, https://github.com/neovim/neovim/pull/10580/commits/65f767a951dca83a8ae57cf77aa072fcaef6e3f2 is not necessary.

**Comment by blueyed** [comment]
@erw7 
Awesome, thanks - that works! :fireworks:

**Comment by blueyed** [comment]
Still a problem with the MSVC_32 job: https://ci.appveyor.com/project/neovim/neovim/builds/26228892/job/vpw5mpgk9h5jo6b1#L6951
Code: https://github.com/blueyed/neovim/blob/d11d74048122d92f94dcf15f205755bc155c2ac8/test/functional/core/job_spec.lua#L709
```
[  ERROR   ] 1 error, listed below:
6952  [  ERROR   ] test/functional\core\job_spec.lua @ 669: jobs jobwait can be called recursively
6953  test/functional\core\job_spec.lua:709: attempt to index local 'r' (a nil value)
6954  
6955  stack traceback:
6956  	test/functional\core\job_spec.lua:709: in function <test/functional\core\job_spec.lua:669>
6957  
6958  
6959   48 SKIPPED TESTS
6960   1 ERROR
6961  -- Tests exited non-zero: 1
6962  -- No output to stderr.
```

**Comment by erw7** [comment]
@blueyed Errors do not occur in my environment, but similar errors can be generated in the following way.

```diff
--- a/test/functional/core/job_spec.lua
+++ b/test/functional/core/job_spec.lua
@@ -702,6 +702,7 @@ describe('jobs', function()
         \ ])
       endfunction
       ]])
+      feed_command('set shell=foo.exe')
       feed_command('call Run()')
       local r
       for i = 10, 1, -1 do
```

```diff
--- a/test/functional/core/job_spec.lua
+++ b/test/functional/core/job_spec.lua
@@ -698,7 +698,7 @@ describe('jobs', function()
         let j.state = 0
         let j.counter = g:counter
         call jobwait([
-        \   jobstart('echo ready; cat -', j),
+        \   jobstart('echo ready; sleep 5; cat -', j),
         \ ])
       endfunction
       ]])
```

If we try the following changes, we may find the cause.

```diff
--- a/test/functional/core/job_spec.lua
+++ b/test/functional/core/job_spec.lua
@@ -702,13 +702,15 @@ describe('jobs', function()
         \ ])
       endfunction
       ]])
-      feed_command('call Run()')
+      command('call Run()')
       local r
       for i = 10, 1, -1 do
-        r = next_msg()
-        eq('job '..i..' closed', r[3][1])
-        r = next_msg()
-        eq('job '..i..' exited', r[3][1])
+        retry(nil, 500, function()
+          r = next_msg()
+          eq('job '..i..' closed', r[3][1])
+          r = next_msg()
+          eq('job '..i..' exited', r[3][1])
+        end)
       end
       eq(10, nvim('eval', 'g:counter'))
     end)
```

**Comment by justinmk** [comment]
@blueyed  skip that case for now, so the rest of this can be merged?

**Comment by blueyed** [comment]
@justinmk 
Ok, following up on that case in https://github.com/neovim/neovim/pull/10607.

**Review by justinmk — APPROVED** [review]

**Pull request merged**